pkg/agent/externalnode/external_node_controller.go

```diff
@@ -0,0 +1,751 @@
+	}
+
+	return attachInfo
+}
+
+func (c *ExternalNodeController) updateOVSPortsData(interfaceConfig *interfacestore.InterfaceConfig, portData *ovsconfig.OVSPortData, eeName string, ips []string) (*interfacestore.InterfaceConfig, error) {
+	attachInfo := map[string]interface{}{
+		ovsExternalIDUplinkName:               portData.ExternalIDs[ovsExternalIDUplinkName],
+		ovsExternalIDUplinkPort:               portData.ExternalIDs[ovsExternalIDUplinkPort],
+		ovsExternalIDEntityName:               eeName,
+		ovsExternalIDEntityNamespace:          portData.ExternalIDs[ovsExternalIDEntityNamespace],
+		ovsExternalIDIPs:                      strings.Join(ips, ipsSplitter),
+		interfacestore.AntreaInterfaceTypeKey: interfacestore.AntreaHost,
+	}
+	err := c.ovsBridgeClient.SetPortExternalIDs(interfaceConfig.InterfaceName, attachInfo)
+	if err != nil {
+		return nil, err
+	}
+	ifIPs := make([]net.IP, 0)
+	for _, ip := range ips {
+		ifIPs = append(ifIPs, net.ParseIP(ip))
+	}
+	iface := &interfacestore.InterfaceConfig{
+		InterfaceName: interfaceConfig.InterfaceName,
+		Type:          interfacestore.ExternalEntityInterface,
+		OVSPortConfig: &interfacestore.OVSPortConfig{
+			PortUUID: interfaceConfig.PortUUID,
+			OFPort:   interfaceConfig.OFPort,
+		},
+		EntityInterfaceConfig: &interfacestore.EntityInterfaceConfig{
+			EntityName:      eeName,
+			EntityNamespace: interfaceConfig.EntityNamespace,
+			UplinkPort: &interfacestore.OVSPortConfig{
+				PortUUID: interfaceConfig.UplinkPort.PortUUID,
+				OFPort:   interfaceConfig.UplinkPort.OFPort,
+			},
+		},
+		IPs: ifIPs,
+	}
+	return iface, nil
+}
+
+func (c *ExternalNodeController) removeOVSPortsAndFlows(interfaceConfig *interfacestore.InterfaceConfig) error {
+	portUUID := interfaceConfig.PortUUID
+	portName := interfaceConfig.InterfaceName
+	if err := c.ofClient.UninstallVMUplinkFlows(portName); err != nil {
+		return err
+	}
+	klog.InfoS(" Removed the flows installed to forward packet between uplinkPort and hostPort.", "hostInterface", portName)
+	hostIFName := interfaceConfig.InterfaceName
+	uplinkIfName := util.GenerateUplinkInterfaceName(portName)
+	uplinkPortID := interfaceConfig.UplinkPort.PortUUID
+	adapterConfig, err := util.GetUplinkConfig(hostIFName)
+	if err != nil {
+		return err
+	}
+	if ovsErr := c.ovsBridgeClient.DeletePort(portUUID); ovsErr != nil {
+		return ovsErr
+	}
+	klog.InfoS("Deleted host port in OVS", "hostInterface", hostIFName)
+	if ovsErr := c.ovsBridgeClient.DeletePort(uplinkPortID); ovsErr != nil {
+		return ovsErr
+	}
+	klog.InfoS("Deleted uplink port in OVS", "uplinkIfName", uplinkIfName)
+	defer func() {
+		// Delete host interface from OVS datapath if it exists.
+		// This is to resolve an issue that OVS fails to remove the interface from datapath. It might happen because the interface
+		// is busy when OVS tries to remove it with the OVSDB interface deletion event.
+		if err := c.ovsctlClient.DeleteDPInterface(hostIFName); err != nil {
```

Comment:
Yes, it is hit in practice. The scene that we hit the error include: 1) the configurations for both uplink and internal port are removed in OVSDB; 2) the interface for the OVS internal port is removed on the host, 3) the uplink interface in OVS datapath is still attached. Since the ports in OVSDB is successfully removed, we hope that the interface in OVS datapath could be removed no matter the later steps are successful or not. So a `defer` is used here.